### ISSUE TITLE
Issue with downloading large blobs

### ISSUE THREAD

**Issue by ddobric** [Azure/azure-sdk-for-net#14955]
**Bug report**
Trying to download a large blob (20GB) by using DownloadToStreamAsync, Getting following exception after approx. 15 min.

> Microsoft.Azure.Storage.StorageException: The operation was canceled.,
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled.,
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, ---> System.IO.IOException: Unable to read data from the transport connection: The I/O operation has been aborted because of either a thread exit or an application request..,
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, ---> System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.,
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, --- End of inner exception stack trace ---,
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)",
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token),
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Security.SslStream.g__InternalFillBufferAsync|215_0[TReadAdapter](TReadAdapter adap, ValueTask1 task, Int32 min, Int32 initial)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory1 buffer)",
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, at System.Net.Http.HttpConnection.ReadAsync(Memory1 destination), 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Http.HttpConnection.ContentLengthReadStream.ReadAsync(Memory1 buffer, CancellationToken cancellationToken)",
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, --- End of inner exception stack trace ---,
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Http.HttpConnection.ContentLengthReadStream.ReadAsync(Memory1 buffer, CancellationToken cancellationToken)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.TimeoutStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.ByteCountingStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Util.TaskExtensions.WithCancellation[T](Task1 task, CancellationToken cancellationToken)",
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Util.AsyncStreamCopier1.StartCopyStreamAsyncHelper(Nullable1 copyLength, Nullable1 maxLength, CancellationToken token)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Util.AsyncStreamCopier1.StartCopyStreamAsync(Nullable1 copyLength, Nullable1 maxLength, CancellationToken cancellationToken)",
> 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Executor.Executor.ExecuteAsync[T](RESTCommand1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, --- End of inner exception stack trace ---, 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Executor.Executor.ExecuteAsync[T](RESTCommand1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)"

**Steps to Reproduce**

Following code fails after 10-15 min. with *TaskCanceledException*.

```

Microsoft.Azure.Storage.CloudStorageAccount storageAccount =
Microsoft.Azure.Storage.CloudStorageAccount.Parse(config.ConnectionString);
            var client = storageAccount.CreateCloudBlobClient();
            var container = client.GetContainerReference(config.ImageSetContainer);
            var cloudBlockBlob = container.GetBlockBlobReference("largefile.zip");

            using (System.IO.Stream file = File.OpenWrite("largefile.zip"))
            {
                await cloudBlockBlob.DownloadToStreamAsync(file, new AccessCondition(), new BlobRequestOptions()
                {
                    RetryPolicy = new Microsoft.Azure.Storage.RetryPolicies.LinearRetry(TimeSpan.FromSeconds(5), 5)
                }, new Microsoft.Azure.Storage.OperationContext(), progressHandler, token);
            }

```

**Expected behavior**
The large blob should be downloaded after a while. A retry should catch temporary network issues.

**Actual behavior**
The code is shown above fails. It breaks downloading. It seems that the stream stops reading for whatever reason.
The exception *ObjectDisposedException*/*TaskCanceledException*/*StorageException* is thrown and no retry is executed. (see trace above)

The code fails in the following method inside of *TimeOutStream.cs*.

```
 public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
        {
            var source = StartTimeout(cancellationToken, out bool dispose);
            try
            {
                return await this.wrappedStream.ReadAsync(buffer, offset, count, source.Token).ConfigureAwait(false);
            }
            // ** ERROR ** 
            catch (ObjectDisposedException)
            {
                source.Token.ThrowIfCancellationRequested();
                throw;
            }
            finally
            {
                StopTimeout(source, dispose);
            }
        }
```

**Environment:**
 - Microsoft.Azure.Storage.Blobs 11.1.5
 - Windows 10 .NET Framework 4.7 and .NET Core 3.1]
 - Executed locally (code shown above downloading the blob from azure storage and the same code executed inside of the web job in the appservice.

### Question:
It seems that the stream just stops reading. Once that happens the operation is cancelled (by default 1:40 min. I tried to increase the NetworkTimeout value to 1hour. The same exception is thrown after 1hour. Is there some way to set some other timeouts in BlobRequestOptions to avoid this issue?
Please note, I have many either blobs 5-10GB size. All works fine instead of this blob which is is 24GB.

**Comment by Ekaterina543** [comment]
> **Bug report**
> 
> Trying to download a large blob (20GB) by using DownloadToStreamAsync, Getting following exception after approx. 15 min.
> 
> 
> 
> > Microsoft.Azure.Storage.StorageException: The operation was canceled.,
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled.,
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, ---> System.IO.IOException: Unable to read data from the transport connection: The I/O operation has been aborted because of either a thread exit or an application request..,
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, ---> System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.,
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, --- End of inner exception stack trace ---,
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)",
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token),
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Security.SslStream.g__InternalFillBufferAsync|215_0[TReadAdapter](TReadAdapter adap, ValueTask1 task, Int32 min, Int32 initial)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory1 buffer)",
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, at System.Net.Http.HttpConnection.ReadAsync(Memory1 destination), 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Http.HttpConnection.ContentLengthReadStream.ReadAsync(Memory1 buffer, CancellationToken cancellationToken)",
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, --- End of inner exception stack trace ---,
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at System.Net.Http.HttpConnection.ContentLengthReadStream.ReadAsync(Memory1 buffer, CancellationToken cancellationToken)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.TimeoutStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.ByteCountingStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Util.TaskExtensions.WithCancellation[T](Task1 task, CancellationToken cancellationToken)",
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Util.AsyncStreamCopier1.StartCopyStreamAsyncHelper(Nullable1 copyLength, Nullable1 maxLength, CancellationToken token)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Util.AsyncStreamCopier1.StartCopyStreamAsync(Nullable1 copyLength, Nullable1 maxLength, CancellationToken cancellationToken)",
> 
> > 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Executor.Executor.ExecuteAsync[T](RESTCommand1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)", 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87, --- End of inner exception stack trace ---, 2020-09-06T21:14:57,Information,trainingjob-test,79216f,637350236977188811,0,8280,87," at Microsoft.Azure.Storage.Core.Executor.Executor.ExecuteAsync[T](RESTCommand1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token)"
> 
> 
> 
> **Steps to Reproduce**
> 
> 
> 
> Following code fails after 10-15 min. with *TaskCanceledException*.
> 
> 
> 
> ```
> 
> 
> 
> Microsoft.Azure.Storage.CloudStorageAccount storageAccount =
> 
> Microsoft.Azure.Storage.CloudStorageAccount.Parse(config.ConnectionString);
> 
>             var client = storageAccount.CreateCloudBlobClient();
> 
>             var container = client.GetContainerReference(config.ImageSetContainer);
> 
>             var cloudBlockBlob = container.GetBlockBlobReference("largefile.zip");
> 
> 
> 
>             using (System.IO.Stream file = File.OpenWrite("largefile.zip"))
> 
>             {
> 
>                 await cloudBlockBlob.DownloadToStreamAsync(file, new AccessCondition(), new BlobRequestOptions()
> 
>                 {
> 
>                     RetryPolicy = new Microsoft.Azure.Storage.RetryPolicies.LinearRetry(TimeSpan.FromSeconds(5), 5)
> 
>                 }, new Microsoft.Azure.Storage.OperationContext(), progressHandler, token);
> 
>             }
> 
> 
> 
> ```
> 
> 
> 
> **Expected behavior**
> 
> The large blob should be downloaded after a while. A retry should catch temporary network issues.
> 
> 
> 
> **Actual behavior**
> 
> The code is shown above fails. It breaks downloading. It seems that the stream stops reading for whatever reason.
> 
> The exception *ObjectDisposedException*/*TaskCanceledException*/*StorageException* is thrown and no retry is executed. (see trace above)
> 
> 
> 
> The code fails in the following method inside of *TimeOutStream.cs*.
> 
> 
> 
> ```
> 
>  public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
> 
>         {
> 
>             var source = StartTimeout(cancellationToken, out bool dispose);
> 
>             try
> 
>             {
> 
>                 return await this.wrappedStream.ReadAsync(buffer, offset, count, source.Token).ConfigureAwait(false);
> 
>             }
> 
>             // ** ERROR ** 
> 
>             catch (ObjectDisposedException)
> 
>             {
> 
>                 source.Token.ThrowIfCancellationRequested();
> 
>                 throw;
> 
>             }
> 
>             finally
> 
>             {
> 
>                 StopTimeout(source, dispose);
> 
>             }
> 
>         }
> 
> ```
> 
> 
> 
> **Environment:**
> 
>  - Microsoft.Azure.Storage.Blobs 11.1.5
> 
>  - Windows 10 .NET Framework 4.7 and .NET Core 3.1]
> 
>  - Executed locally (code shown above downloading the blob from azure storage and the same code executed inside of the web job in the appservice.
> 
> 
> 
> ### Question:
> 
> It seems that the stream just stops reading. Once that happens the operation is cancelled (by default 1:40 min. I tried to increase the NetworkTimeout value to 1hour. The same exception is thrown after 1hour. Is there some way to set some other timeouts in BlobRequestOptions to avoid this issue?
> 
> Please note, I have many either blobs 5-10GB size. All works fine instead of this blob which is is 24GB.


name : Import Secrets 
      использует : hashicorp / vault-action 
      с :
         url : https://vault-enterprise.mycompany.com:8200 
        method : token 
        caCertificate : $ {{secrets.VAULTCA}} 
        токен : $ {{secrets.VaultToken}} 
        пространство имен : ns1 
        secrets : | 
            секрет / ci / aws accessKey | AWS_ACCESS_KEY_ID; 
            secret / ci / aws secretKey | AWS_SECRET_ACCESS_KEY; 
            секрет / ci npm_token

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team best able to assist.

**Comment by ddobric** [comment]
Thanks @jsquire.

I have found how the issue can be fixed. By commenting the line [172 ](https://github.com/Azure/azure-storage-net/blob/25e73e3c53a15e7cb4c2d1b2170065bf4bfa73dc/Lib/ClassLibraryCommon/Core/TimeoutStream.cs)

![image](https://user-images.githubusercontent.com/1756871/92563365-c5b19100-f277-11ea-9f41-033e51feb9fc.png)

However, the implementation looks as a feature and I'm not sure that commenting that out is the correct  way.  The cancelation token consists of two cancelations. One of them is bound to the NetworkTimeout, which causes that issue. So, the final solution must distinguish between two cancelation tokens inside of ReadAsync and prevent throwing of ObjectDisposedException when network timeout expires.
With that, enabled retry mechanism will retry the operation and all will work find.

To recap, the committed version does not work. with large files, because transient errors will not be handled throwing exception in the line 172. Not going to create PR, because the right solution requires more changes inside of ReadAsync (i.e.: split of linked tokens).

Hope this helps.

Damir

**Comment by amnguye** [comment]
Hi,

This is a v11 issue you submitted, not a v12 issue. I recommend that you either reopen the original issue you had in the old library repo or you upgrade to the v12 library and see if any of the fixes or enhancements we have done in the v12 library resolve your issue.

Please reactivate if this is a v12 issue.

**Comment by ddobric** [comment]
@amnguye this is not a v11 issue only. The same issue is in v12 too.

### Exception:
   at Microsoft.Azure.Storage.Core.Executor.Executor.<ExecuteAsync>d__1`1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Daenet.ProductIdentification.Core.Storage.AzureStorageProvider.<DownloadBlobFile>d__43.MoveNext() in ...

### Inner Exception 1:
   at System.Net.Http.HttpConnection.ContentLengthReadStream.<ReadAsync>d__3.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Storage.Core.TimeoutStream.<ReadAsync>d__34.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Storage.Core.ByteCountingStream.<ReadAsync>d__27.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Storage.Core.Util.TaskExtensions.<WithCancellation>d__0`1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at Microsoft.Azure.Storage.Core.Util.AsyncStreamCopier`1.<StartCopyStreamAsyncHelper>d__16.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Storage.Core.Util.AsyncStreamCopier`1.<StartCopyStreamAsync>d__15.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at Microsoft.Azure.Storage.Core.Executor.Executor.<ExecuteAsync>d__1`1.MoveNext()

### Inner Exception 2
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token)
   at System.Net.Security.SslStream.<<FillBufferAsync>g__InternalFillBufferAsync|215_0>d`1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Net.Security.SslStream.<ReadAsyncInternal>d__214`1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpConnection.<ReadAsync>d__90.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpConnection.ContentLengthReadStream.<ReadAsync>d__3.MoveNext()

### Inner Exception 3:
The I/O operation has been aborted because of either a thread exit or an application request.
